src/qiboml/models/pytorch.py

```diff
@@ -0,0 +1,63 @@
+"""Torch interface to qiboml layers"""
+
+from dataclasses import dataclass
+
+import numpy as np
+import torch
+from qibo.config import raise_error
+
+import qiboml.models.encoding_decoding as ed
+from qiboml.models.abstract import QuantumCircuitLayer
+
+
+@dataclass
+class QuantumModel(torch.nn.Module):
+
+    layers: list[QuantumCircuitLayer]
+
+    def __post_init__(self):
+        super().__init__()
+        for layer in self.layers[1:]:
+            if layer.circuit.nqubits != self.nqubits:
+                raise_error(
+                    RuntimeError,
+                    f"Layer \n{layer}\n has {layer.circuit.nqubits} qubits, but {self.nqubits} qubits was expected.",
+                )
+            if layer.backend.name != self.backend.name:
+                raise_error(
+                    RuntimeError,
+                    f"Layer \n{layer}\n is using {layer.backend} backend, but {self.backend} backend was expected.",
+                )
+        for layer in self.layers:
+            if len(layer.circuit.get_parameters()) > 0:
+                self.register_parameter(
+                    layer.__class__.__name__,
+                    torch.nn.Parameter(torch.as_tensor(layer.circuit.get_parameters())),
+                )
+        if not isinstance(self.layers[-1], ed.QuantumDecodingLayer):
+            raise_error(
+                RuntimeError,
+                f"The last layer has to be a `QuantumDecodinglayer`, but is {self.layers[-1]}",
+            )
+
+    def forward(self, x: torch.Tensor):
+        if self.backend.name != "pytorch":
```

Comment:
Any reason to use PyTorch with a non-Pytorch backend?

Reply:
Similarly to the comment above, this is the `pytorch` interface, which consumes `torch.tensors` but this is separated from the backend you are using. If the backend is the `PyTorchBackend`, then you don't need to do anything, otherwise you have to cast to the appropriate type.

Reply:
Yes indeed, but since there is currently no implementation of the `evaluate` function, as gradients are still not in main yet, I had to take care of that here in order to test. As I said in one of the previous meetings, this object is probably going to change when we incorporate the auto differentiation, as pytorch wants you to write a custom autograd function which is probably going to perform what is done here in the `forward`.

Reply:
Ok, maybe it's worth to commit a bit of time to the Pytorch implementation, since (in principle) `expectation()` is in `main`:
https://github.com/qiboteam/qiboml/blob/0e906eef4117da995e3897c7b50c92c8bf6e6a1e/src/qiboml/operations/expectation.py#L12-L19
(sorry, I messed up, it's not `evaluate()`)

Working on top of that, should make the model structure more compatible.

In any case, if you prefer doing it in a different PR I'm not completely against. But you might end up reverting there part of what is done here.